cmd/restic/cmd_restore.go

```diff
@@ -171,9 +178,12 @@ func runRestore(ctx context.Context, opts RestoreOptions, gopts GlobalOptions,
 			matched, childMayMatch := includeFn(item)
 			selectedForRestore = selectedForRestore || matched
 			childMayBeSelected = childMayBeSelected || childMayMatch
-		}
+			childMayBeSelected = childMayBeSelected && node.Type == "dir"
 
-		childMayBeSelected = childMayBeSelected && node.Type == "dir"
+			if selectedForRestore || childMayBeSelected {
```

Comment:
I think that is wrong. If the first `includeFn` returns `false, true` (matched, childMayMatch) for a directory, then the `if` would exit the loop right now. However, a second `includeFn` could still return `true, true`. 

To handle that case correctly, the loop should look like the following:
```Go
for _, includeFn := range includePatternFns {
			matched, childMayMatch := includeFn(item)
			selectedForRestore = selectedForRestore || matched
			childMayBeSelected = childMayBeSelected || childMayMatch

			if selectedForRestore && childMayBeSelected {
				break
			}
		}
		childMayBeSelected = childMayBeSelected && node.Type == "dir"
```

Reply:
Done.